dev/com.ibm.ws.microprofile.faulttolerance.2.0/src/com/ibm/ws/microprofile/faulttolerance20/state/impl/RetryStateImpl.java

```diff
@@ -0,0 +1,146 @@
+/*******************************************************************************
+ * Copyright (c) 2018 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.microprofile.faulttolerance20.state.impl;
+
+import java.time.Duration;
+import java.util.PrimitiveIterator;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+
+import com.ibm.ws.microprofile.faulttolerance.spi.RetryPolicy;
+import com.ibm.ws.microprofile.faulttolerance20.impl.MethodResult;
+import com.ibm.ws.microprofile.faulttolerance20.state.RetryState;
+
+public class RetryStateImpl implements RetryState {
+
+    private static final Random random = new Random();
+
+    private final RetryPolicy policy;
+    private final long delayMillis;
+    private final long jitterMillis;
+    private final PrimitiveIterator.OfLong jitterStream;
+    private int attempts = 0;
+    private long startNanos;
+
+    public RetryStateImpl(RetryPolicy policy) {
+        this.policy = policy;
+        delayMillis = policy.getDelay().toMillis();
+        jitterMillis = policy.getJitter().toMillis();
+
+        if (jitterMillis != 0L) {
+            jitterStream = random.longs(delayMillis - jitterMillis, delayMillis + jitterMillis).iterator();
```

Comment:
comment

Reply:
I've done some refactoring here and added some comments which hopefully makes it clearer.

Reply:
Done